Edit dart pub token page (#5717)
dart-lang/site-www: src/content/tools/pub/cmd/pub-token.md

description: Manage authentication tokens for package repositories.
---

The `dart pub token` subcommand manages a store of secret tokens to
authenticate against third-party servers when [publishing](pub-lish) packages
and [retrieving](pub-get) dependencies.

It stores these tokens in a [user-wide config directory][config-dir].

> It has three subcommands: `add`, `list` and `remove`.

Reviewer: suggestion (lite): This gets tricky with avoiding passive voice, but could we say something besides "It" again here (the previous sentence also starts with "It")

- `dart pub token` has three subcommands....
- The subcommand has three subcommands of its own...

I don't have any good ideas, just a thought.

> It has three subcommands: `add`, `list` and `remove`.

Author: **suggestion:** 
```suggestion
This command consists of three subcommands: `add`, `list` and `remove`.
```

> This command consists of three subcommands: `add`, `list` and `remove`.

Reviewer: ```suggestion
The command hs of three subcommands: `add`, `list` and `remove`.
```
"consists" isn't really the right word, "has" or "includes"

> Consider a scenario when you have a [dependency](/tools/pub/dependencies)

Reviewer: ```suggestion
Consider a scenario where you have a [dependency](/tools/pub/dependencies)
```

Go to https://some-package-repo.com and log in to obtain your token. 
```

> In this last message, the final line gives you the server that can help you
> obtain a token. Not all servers provide such this message.

Reviewer: ```suggestion
The server might include a message in the final line of the prompt
that can help you obtain a token, shown in the example above.
Not all servers provide this message.
```

## Add a new credential

> To create a new credential, invoke `dart pub token add`.

Reviewer: ```suggestion
To add a new credential, run the command `dart pub token add`.
```

The command allows the user to _add_ a credential to the repository, but doesn't create a credential for them... It's unclear from this doc where the user gets the credential from, but I think they just make one up themselves when prompted by the `add` command

### Add a credential for the current session

> At the prompt, type the credential on the command line (`stdin`).

Reviewer: ```suggestion
At the prompt, type the credential on the command line.
```

Since this is covered in the paragraph following the example prompt, I don't think it needs to be mentioned here.

```

:::note

> The token takes input on `stdin` rather than as a command line option

Reviewer: ```suggestion
The token takes input over `stdin` rather than as a command line option
```

to keep the token out of the shell history.
:::

### Add a credential for all sessions

> In a script, you can store the secret in an environment variable.

Reviewer: I think the more critical info for this page/topic should come first in the section, like this:

````suggestion
In a script, you can store the secret token in an environment variable.
Run the following command:

```console
dart pub token add <hosted-url> --env-var <ENV_VAR_NAME>
```

This causes `dart pub get` to read the data stored in `$TOKEN_VAR` then
use it as the authentication token.

```console
$ dart pub token add https://other-package-repo.com/ --env-var TOKEN_VAR
Requests to "https://other-package-repo.com/" will now be authenticated using the secret token stored in the environment variable "TOKEN_VAR".
```
````

Reviewer: It [looks like](https://dart-dev--pr5717-fix-3729-y5ryp88s.web.app/tools/pub/cmd/pub-token#add-a-credential-for-all-sessions) you removed all of this content (the syntax/sample code) from the page, accident or intentional?

> If you choose to store your secret in an environment variable,
> find a way to hide the secret from your shell history.
> To explore one way of doing this, consult [this post on Medium][zsh-post].

Reviewer: ```suggestion
If you choose to store your token in an environment variable,
[find a way][zsh-post] to hide the token from your shell history.
```

I think it's odd to call out a medium post that wasn't written by us. Suggestion deemphasizes it.

> If you add a local environment variable, you need to restart any open

Reviewer: ```suggestion
If you add an environment variable, you need to restart any open
```

"local env variable" is redundant

consoles to enable that new variable.

> Most CI environments can inject secrets into an environment variable.
> To learn how, consult documentation for [GitHub Actions][]] or
> [GitLab][] as examples.

Reviewer: ```suggestion
Most CI environments can inject tokens into an environment variable.
To learn how, consult documentation for [GitHub Actions][] or
[GitLab][] as examples.
```

Author: **comment:** Agreed, but CI systems usually call them secrets. I know, it's just synonym city here.

Reviewer: You missed the extra bracket after GitHub Actions

[GitHub Actions]: https://docs.github.com/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow
[GitLab]: https://docs.gitlab.com/ee/ci/secrets/
[zsh-post]: https://medium.com/@prasincs/hiding-secret-keys-from-shell-history-part-1-5875eb5556cc

> To use an environment variable as a token, invoke:

Reviewer: ```suggestion
```

moved up in the section

> ```console
> dart pub token add <hosted-url> --env-var <ENV_VAR_NAME>
> ```

Reviewer: ```suggestion
```

moved up in the section

> This causes `dart pub get` to read the data stored in `$TOKEN_VAR` then
> use it as the authentication token.

Reviewer: ```suggestion
```

moved up in the section

> ```console
> $ dart pub token add https://other-package-repo.com/ --env-var TOKEN_VAR
> Requests to "https://other-package-repo.com/" will now be authenticated using the secret token stored in the environment variable "TOKEN_VAR".
> ```

Reviewer: ```suggestion
```

moved up in the section

## Return a list of credentials

> To see a list of all active credentials, invoke `dart pub token list`:

Reviewer: ```suggestion
To see a list of all active credentials, run the command `dart pub token list`:
```

https://other-package-repo.com/
```

## Remove one or more credentials

> To remove a single token, invoke `dart pub token remove`:

Reviewer: ```suggestion
To remove a single token, run the command `dart pub token remove`:
```

```console
$ dart pub token remove https://other-package-repo.com
Removed secret token for package repository: https://other-package-repo.com
```

> To remove all tokens, invoke the command with the `remove --all` option:

Reviewer: ```suggestion
To remove all tokens, run the same with the `remove --all` option:
```
